src/Resolution.ts

```diff
@@ -732,24 +772,60 @@ export default class Resolution {
     return records[newRecord] || records[oldRecord];
   }
 
-  private getNamingMethod(domain: string): NamingService | undefined {
-    return this.serviceMap[findNamingServiceName(domain)];
+  private getNamingMethods(domain: string): NamingService[] {
+    const method = this.serviceMap[findNamingServiceName(domain)];
+    if (!method) {
+      return [];
+    } else if (method === this.serviceMap.ZNS) {
+      return [this.serviceMap.UNS, method];
+    } else {
+      return [method];
+    }
   }
 
-  private getNamingMethodOrThrow(domain: string): NamingService {
-    const method = this.getNamingMethod(domain);
-    if (!method) {
+  private async callForDomain<F extends keyof NamingService>(
+    domain: string,
+    methods: NamingService[],
+    func: F,
+    args: Parameters<NamingService[F]>,
```

Comment:
What if we pass a lambda that should be called for each naming service instead of the raw parameters?
```
private async callForDomain<TReturn>(
    methods: NamingService[],
    func: (NamingService) : Promise<TReturn>
 ): Promise<TReturn> {
    const servicePromises = methods.map((method) =>
       func(method)
    );
    ...
}
```
It seems like it should work and make the typing easier, but I'm not sure.